### ISSUE TITLE
Uploading to cloud platforms: GCP

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#147]
This is part of https://github.com/coreos/fedora-coreos-tracker/issues/146 and tracks the work/discussion around uploading to GCP.

**Comment by bgilbert** [comment]
I think this involves the following:
- [ ] Have GCP create a project to host the images.  That project is traditionally called `<something>-cloud`.  We should decide whether the project should be FCOS-specific or Fedora-wide.  Our GCP contacts would like to meet with someone from releng to start setting this up.
- [ ] Adapt plume for FCOS.
- [ ] After testing, have GCP mark the project public.

**Comment by dustymabe** [comment]
related: https://github.com/coreos/coreos-assembler/issues/493

**Comment by bgilbert** [comment]
[Infra ticket](https://pagure.io/fedora-infrastructure/issue/8281) to create the requisite GCP projects.

**Comment by cgwalters** [comment]
I'm working on some GCP bits for RHCOS (see e.g. https://github.com/openshift/installer/pull/2921 ) and not having FCOS there inhibits doing things upstream first.

**Comment by dghubble** [comment]
With manually uploaded GCP images, zincati checks for updates and reboots promptly (speedy!) after boot which has some interesting cascading affect on interrupting inital cluster [bootstrapping](https://github.com/poseidon/typhoon/pull/687) that I noticed today. Being able to use a latest GCP channel image would have a nice side-benefit of (mostly) clearing up immediate reboot issues.

**Comment by cgwalters** [comment]
@dghubble slightly related https://github.com/coreos/zincati/pull/251
But you probably need to be using https://github.com/coreos/airlock or an equivalent if you aren't.

**Comment by LorbusChris** [comment]
I think also related here is https://github.com/coreos/fedora-coreos-tracker/issues/392

In OKD we're working around this by adding a `etc/zincati/config.d/90-disable-feature.toml` that explicitly disables Zincati updates with Ignition: https://github.com/openshift/installer/blob/fcos/data/data/bootstrap/files/etc/zincati/config.d/90-disable-feature.toml

**Comment by bgilbert** [comment]
@LorbusChris That approach makes sense when the cluster has its own mechanism for updating nodes.  In general, though, we shouldn't encourage users to disable updates.

**Comment by LorbusChris** [comment]
Indeed, we shouldn't encourage that in general. My thinking was that in this specific case of a (probably short-lived) bootstrap host it may make some sense :)

**Comment by dghubble** [comment]
Thanks for the suggestions. For now I'm content to follow this issue for GCP uploads. Manually uploading new images or sometimes having to retry the bootstrap in this one case is ok to me as I'd like to keep auto-updates enabled (its a master/controller, rather than a toss-away like OKD uses I think).

Mainly just highlighting that stale manual images can cause these reboots (e.g. new instance creation, after GCE preemption) that wouldn't otherwise be needed, until the channel is in place.

**Comment by lucab** [comment]
@dghubble good feedback points on https://github.com/poseidon/typhoon/pull/687. I have a few more things for you:
 1. Zincati only runs [after](https://github.com/coreos/zincati/blob/v0.0.9/dist/systemd/system/zincati.service#L10) `boot-complete.target`. If your bootstrapping flow needs to happen before any update, you can delay reaching that target until done.
 2. if you are bootstrapping from scratch and can't afford some [lock-managing service](https://github.com/coreos/zincati/blob/master/docs/usage/updates-strategy.md#lock-based-strategy) around, it sounds like you may want something like https://github.com/coreos/zincati/issues/245 implemented
 3. similar to the above, path-activation may be interesting too but I'm unsure on how to plug the default-on and Ignition disabling into that
 3. in general some deterministic trigger is preferred over an arbitrary timer/delay. Both the timer-activation and the SSH-delay would just make races harder to detect.

**Comment by dghubble** [comment]
Thanks @lucab!  I think I'd wanna aim for the lock managing approach personally - a local file strategy could help in a pinch here (for bootstrap) and a later an airlock atop the cluster could get (roughly) one at a time node updates (I don't have strong consistency needs). I looked at airlock's etcd strategy, but I'm reluctant to give it access policy-wise or deal with giving it a TLS client cert, etc. I'd picture airlock maybe running as a pod with rbac to write a configmap or annotation, some bit of scratch space for locked yes/no. airlock is looking promising and simpler than cluo was, avoiding the coordinator/daemonset.

For 1, I'd worry about blocking the target on future reboots (bootstrap is once per cluster lifetime). 3 yeah is kinda , and agreed that 4 may not be the best for this situation.

Maybe I'll move this to a (new?) ticket in airlock so I don't hijack.

**Comment by dustymabe** [comment]
Some updates...

> I think this involves the following:
> 
> * [ ]  Have GCP create a project to host the images.  That project is traditionally called `<something>-cloud`.  We should decide whether the project should be FCOS-specific or Fedora-wide.  Our GCP contacts would like to meet with someone from releng to start setting this up.

We have a `fedora-coreos-cloud` project now.

>
> * [ ]  Adapt plume for FCOS.

We adapted `ore` within COSA to support uploading to GCP and doing what we need.

- https://github.com/coreos/coreos-assembler/pull/1333
- https://github.com/coreos/coreos-assembler/pull/1328
- https://github.com/coreos/coreos-assembler/pull/1322
- https://github.com/coreos/coreos-assembler/pull/1319

We modified the pipeline to start doing the upload:

- https://github.com/coreos/fedora-coreos-pipeline/pull/219
- https://github.com/coreos/fedora-coreos-pipeline/pull/220

>
> * [ ]  After testing, have GCP mark the project public.

GCP is working on making the `fedora-coreos-cloud` project public. In the meantime I've individually marked the images from our latest testing and stable releases as public so you should be able to use them today:


```
# gcloud compute images list --project fedora-coreos-cloud --no-standard-images --show-deprecated 
NAME                                       PROJECT              FAMILY                       DEPRECATED  STATUS
fedora-coreos-31-20200323-3-2-gcp-x86-64   fedora-coreos-cloud  fedora-coreos-stable                     READY
fedora-coreos-31-20200407-2-2-gcp-x86-64   fedora-coreos-cloud  fedora-coreos-testing                    READY
```

**Comment by vrutkovs** [comment]
Is there a bucket with `.tar.gz` artifact uploaded so that OKD could reuse it?

**Comment by dustymabe** [comment]
the url for the tar.gz within GCP is in the meta.json. My understanding is that it's accessible to any authenticated users. Previously I had considered the file in the image bucket to be purely ephemeral (i.e. only needing to exist for the image import to happen), but I've recently learned of openshift creating images on the fly during install. I'd like to understand why we need to do that if an image is available publicly (and globally) already? Different guestOSfeatures? Encryption?

**Comment by cgwalters** [comment]
>  but I've recently learned of openshift creating images on the fly during install. I'd like to understand why we need to do that if an image is available publicly (and globally) already? Different guestOSfeatures? Encryption?

The idea broadly speaking was that RHEL CoreOS is an implementation detail of OpenShift.  Hence, there wasn't a desire to publish pre-built images globally.  This might change at some point.

It would seem reasonable to me though to change openshift-install to support directly using a public GCP image if it's available, it'd just require some conditionals in the logic for terraform.

**Comment by dustymabe** [comment]
> > but I've recently learned of openshift creating images on the fly during install. I'd like to understand why we need to do that if an image is available publicly (and globally) already? Different guestOSfeatures? Encryption?
> 
> The idea broadly speaking was that RHEL CoreOS is an implementation detail of OpenShift. Hence, there wasn't a desire to publish pre-built images globally. This might change at some point.

Yeah I think that makes sense. Since we are publishing images already for Fedora CoreOS I think it makes sense to do what you suggest below 👇🏼 for Fedora CoreOS/OKD.
> 
> It would seem reasonable to me though to change openshift-install to support directly using a public GCP image if it's available, it'd just require some conditionals in the logic for terraform.

**Comment by dustymabe** [comment]
remaining items here:

- [ ] update mantle to use new image update API so we can create images not in an image family and then later add them to an image family
- [ ] attach "licenses" to images when they get created (this allows for viewing how many instances are created on each stream)
- [ ] documentation page for GCP https://github.com/coreos/fedora-coreos-tracker/issues/475
- [ ] change image size to be 10GiB or greater (known to have suboptimal I/O performance with <10G)

**Comment by dustymabe** [comment]
updated https://github.com/coreos/fedora-coreos-tracker/issues/147#issuecomment-625016777 with current status.

**Comment by lucab** [comment]
@dghubble yes, feel free to move to another ticket or a [forum](https://discussion.fedoraproject.org/c/server/coreos) discussion. From your last reply, it looks like a volatile fragment in `/run` to use https://github.com/coreos/zincati/issues/245 on first-boot only could be a good solution.

**Comment by dustymabe** [comment]
updated https://github.com/coreos/fedora-coreos-tracker/issues/147#issuecomment-625016777 with current status.

**Comment by dghubble** [comment]
I'm using the new uploaded GCP images instead of manually uploading. Many thanks for these!

```
# Terraform example
# Fedora CoreOS most recent image from stream
data "google_compute_image" "fedora-coreos" {
  project = "fedora-coreos-cloud"
  family  = "fedora-coreos-${var.os_stream}"
}
```

**Comment by dustymabe** [comment]
ok all things in https://github.com/coreos/fedora-coreos-tracker/issues/147#issuecomment-625016777 are now done or broken out into other tickets. I also created a new ticket for us getting this plumbed through to our download page: https://github.com/coreos/fedora-coreos-tracker/issues/494


 I'm going to mark this as done. It was a long journey.
